cluster/engine.go

```diff
@@ -739,31 +727,67 @@ func (e *Engine) refreshVolume(IDOrName string) error {
 // true, each container will be inspected.
 // FIXME: unexport this method after mesos scheduler stops using it directly
 func (e *Engine) RefreshContainers(full bool) error {
-	e.Lock()
-	defer e.Unlock()
 	opts := types.ContainerListOptions{
 		All:  true,
 		Size: false,
 	}
+	// Get the current list of containers and cache it
+	e.RLock()
+	currentContainerIDs := make([]string, 0, len(e.containers))
+	for k := range e.containers {
+		currentContainerIDs = append(currentContainerIDs, k)
+	}
+	e.RUnlock()
+
 	ctx, cancel := context.WithTimeout(context.TODO(), requestTimeout)
 	defer cancel()
 	containers, err := e.apiClient.ContainerList(ctx, opts)
-	e.checkConnectionErr(err, true)
+	e.CheckConnectionErr(err)
 	if err != nil {
 		return err
 	}
 
+	var missingContainerIDs []string
```

Comment:
@dani-docker can you add a comment here just briefly describing the logic and reasoning for doing it this way? It's fairly clear in the context of the PR, but it would help in the future.